src/main/java/com/adobe/cq/commerce/graphql/client/impl/CacheKey.java

```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+ *
+ *    Copyright 2020 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+package com.adobe.cq.commerce.graphql.client.impl;
+
+import java.util.Objects;
+
+import com.adobe.cq.commerce.graphql.client.GraphqlRequest;
+import com.adobe.cq.commerce.graphql.client.RequestOptions;
+
+public class CacheKey {
```

Comment:
Any thoughts on the possible performance implications of this heavy cache key?

Reply:
I did a few tests, it took about 0.3ms to compute the `hash` once the classes were loaded once, and a few milliseconds for the `equals` method. So in case there isn't any cache hit, the penalty would typically be less than 1ms because it only needs to compute the `hash` (and exceptionally call `equals` in case of hash collision), and if there is a cache hit, it would take a few ms instead of a round-trip networking time. 

Note that the key is as "heavy" as it needs to be: we have to check the query and the headers, and I allowed headers to be in different order because ordering of headers doesn't matter.

Reply:
Depending on cache configuration and data volume equals() may easily be invoked more than once both for cache misses and cache hits.